src/Pact/Native/Trans.hs

```diff
@@ -133,21 +154,45 @@ trans_sqrt i x = go
 
 liftUnDec
   :: HasInfo i => i
-  -> (Decimal -> TransResult Decimal)
+  -> (TransGasParams -> Decimal -> TransResult (Decimal, Word64))
   -> Decimal -> Eval e Decimal
-liftUnDec i f a = checkTransResult i (f a)
+liftUnDec i f a = do
+  gc <- getTransGasParams
+  checkGasTransResult i (f gc a)
 
 liftBinDec
   :: HasInfo i => i
-  -> (Decimal -> Decimal -> TransResult Decimal)
+  -> (TransGasParams -> Decimal -> Decimal -> TransResult (Decimal, Word64))
   -> Decimal -> Decimal -> Eval e Decimal
-liftBinDec i f a b = checkTransResult i (f a b)
+liftBinDec i f a b = do
+  gc <- getTransGasParams
+  checkGasTransResult i (f gc a b)
 
 liftBinInt
   :: HasInfo i => i
-  -> (Decimal -> Decimal -> TransResult Decimal)
+  -> (TransGasParams -> Decimal -> Decimal -> TransResult (Decimal, Word64))
   -> Integer -> Integer -> Eval e Integer
-liftBinInt i f a b = d2Int <$> checkTransResult i (int2D a `f` int2D b)
+liftBinInt i f a b = do
+  gc <- getTransGasParams
+  d2Int <$> checkGasTransResult i (f gc (int2D a) (int2D b))
+
+chargePactArith :: GasArithOp a b c d -> Arith p r ()
+chargePactArith (GasArithOp _ l r) = 
+  chargeArithGas (fromIntegral (decimalCost l + decimalCost r + 1))
+
+decimalCost :: NDec.Decimal a b -> Gas
+decimalCost (NDec.Num _sign coeff _exp) 
+  | intValue < threshold = 0
+  | otherwise = 
+    let !nbytes = (I# (IntLog.integerLog2# intValue) + 1) `quot` 8
+    in fromIntegral (nbytes * nbytes `quot` 100)
+  where
+  threshold :: Integer
+  threshold = (10 :: Integer) ^ (80 :: Integer)
```

Comment:
Basically, decimal with a coefficient `< 10^80` is free of charge?  Shouldn't our accounting incorporate the exponent? 

Also, shouldn't we account for everything (gas `> 0`), including the case `intValue < threshold`?

Reply:
above 10^80 it's a penalty. We do charge 1 gas for integers larger than this. It will become more granular in the future.

We don't account for the exponent atm, but there's not all that much arithmetic around it.